rust/sdk/src/client.rs

```diff
@@ -457,6 +458,684 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+    ///             },
+    ///             PublishEvent {
+    ///                 tags: tags!("temperature", "sensor:temp-sensor2"),
+    ///                 payload: Payload::compact(&serde_json::json!({ "temperature": 27 })).unwrap(),
+    ///             },
+    ///         ])
+    ///         .await
+    ///         .unwrap()
+    /// }
+    /// ```
+    pub fn events<E: IntoIterator<Item = impl Into<PublishEvent>>>(mut self, events: E) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.data.extend(events.into_iter().map(Into::into));
+            return self;
+        }
+        panic!("Calling Publish::events after polling.");
+    }
+}
+
+impl<'a> Future for Publish<'a> {
+    type Output = anyhow::Result<PublishResponse>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        let this = self.get_mut();
+        loop {
+            *this = match replace(this, Publish::Void) {
+                Publish::Initial { client, request } => {
+                    let publish_response = async move {
+                        let publish_url = client.events_url("publish");
+                        let response = client.do_request(|c| c.post(publish_url).json(&request)).await?;
+                        let body = response.bytes().await?;
+                        Ok(serde_json::from_slice::<PublishResponse>(&body)?)
+                    };
+                    Publish::Pending(publish_response.boxed())
+                }
+                Publish::Pending(mut publish_response_future) => {
+                    let polled = publish_response_future.poll_unpin(cx);
+                    if polled.is_pending() {
+                        *this = Publish::Pending(publish_response_future);
+                    }
+                    return polled;
+                }
+                Publish::Void => panic!("Polling a terminated Publish future"),
+            };
+        }
+    }
+}
+
+impl<'a> FusedFuture for Publish<'a> {
+    fn is_terminated(&self) -> bool {
+        matches!(self, Publish::Void)
+    }
+}
+
+/// Request builder for queries.
+///
+/// Warning: [`Query`] implements the [`Future`] trait, as such it can be polled.
+/// Calling _any_ [`Query`] function after polling will result in a panic!
+pub enum Query<'a> {
+    Initial { client: &'a Ax, request: QueryRequest },
+    Pending(BoxFuture<'a, anyhow::Result<BoxStream<'static, QueryResponse>>>),
+    Void,
+}
+
+impl<'a> Query<'a> {
+    fn new<Q: Into<String>>(client: &'a Ax, query: Q) -> Self {
+        Self::Initial {
+            client,
+            request: QueryRequest {
+                query: query.into(),
+                lower_bound: Some(OffsetMap::empty()),
+                upper_bound: None,
+                order: Order::Asc,
+            },
+        }
+    }
+
+    /// Add a (inclusive) lower bound to the query.
```

Comment:
the lower bound is exclusive (and the upper bound inclusive) because that is much simpler to track than the other way around: the lower bound defines what you have already seen and don’t want to see again

Reply:
Indeed, I had an extra print in my test example and got confused. Nice catch!